File: python/LinkedList.py


class Node:
    data: int
    next: Node

    def __init__(self, data, next=None):
        self.data = data
        self.next = next


class LinkedList:
    head: Node
    def __init__(self):
        self.head = None

    # function to insert at the start of the list
    """
    we first check if the list is empty
    if the list is empty than we make the insererted element the head and return the funtion
    if the list is not empty than we make a new node, and set the head of the list as new node next pointer
    then we set the new node as the head of the list
    """
    def insertAtStart(self, data):
        # check if list is empty
        if self.head == None:
            new_node = Node(data=data)
            # set new node as the head
            self.head = new_node
            print("Node has been inserted at the start")
            return new_node
        
        new_node = Node(data=data)
        # set new nodes next pointer to the current head, moving the head to the second position in the list
        new_node.next = self.head
        # set the new node as the new head of the list
        self.head = new_node
        print("Node has been inserted, head updated")
        return new_node
    
    # function to insert at the end of the list
    """
    we first check if the list is empty
    if list is not empty we then traverse to the last node
    we set the last node's next pointer to the new node
    we set new node next pointer to null since it is at the end of the list
    """
    def insertAtEnd(self, data):
        # check if the list is empty
        if self.head == None:
            new_node = Node(data=data)
            self.head = new_node
            print("The list was empty! Element added as head!")
            return
        
        new_node = Node(data=data)
        # traverse the list until we are at last node
        current = self.head
        while current.next is not None:
            current = current.next
        
        current.next = new_node;
        print("Node inserted at the end of the list!")
        return

    # insert at position function
    """
    handles node insertion at any given position
    does not handle out of bounds!
    we first check if the list is empty
    we then check if the position given is 0 and call insertAtStart function in that case
    we then traverse to the node at pos-1
    we then set the next pointers according to new node, new_node.next = pos-1.next and then pos-1.next to new_node
    """
    def insertAtPosition(self, data, pos):
        # check if the list is empty
        if self.head == None:
            new_node = Node(data=data)
            self.head = new_node
            print("The List was empty! Element added as head!")
            return
        
        # handle position 0
        if pos == 0:
            self.insertAtStart(data=data)
            return
        
        current = self.head
        current_pos = 0
        while current_pos != pos-1:
            current_pos += 1
            current = current.next
        
        new_node = Node(data)
        new_node.next = current.next
        current.next = new_node
        print(f"Node Inserted at position {pos}")

    # function to traverse the list
    """
    first we check if the list is empty
    if the list is not empty then we set the current pointer to the head of the list
    we print the data of the current node and move current pointer to the nodes next pointer
    we do this until current pointer is not null
    """
    def traverse(self):
        # check if the list is empty
        if (self.head == None):
            print("The list is empty nothing to traverse")
            return
        
        current = self.head
        while current is not None:
            print(current.data)
            current = current.next
        


list = LinkedList()
list.insertAtStart(10)
list.insertAtStart(15)
list.insertAtEnd(5)
list.insertAtPosition(99, 1)
list.traverse()

File: c/linkedlist.c
// implementation of linked list in c

#include <stdio.h>
#include <stdlib.h>

// Node struct

struct Node {
    int data;
    struct Node *next;
};

// Linked List Struct

struct LinkedList {
    int length;
    struct Node *head;
};

struct Node *createNode(int data) {
    struct Node *newNode = malloc(sizeof(struct Node));

    if (newNode == NULL) {
        return NULL;
    }

    newNode->data = data;
    newNode->next = NULL;
    return newNode;
}

struct LinkedList *createList() {
    struct LinkedList *list = malloc(sizeof(struct LinkedList));

    if (list == NULL) {
        return NULL;
    }

    list->length = 0;
    list->head = NULL;
    return list;
}

void appendAtStart(struct LinkedList *list, struct Node *node) {
    // check if the list is empty
    if (list->length == 0) {
        list->head = node;
        list->length++;
        printf("Node Appended!\n");
        return;
    }

    // if list is not empty
    node->next = list->head;
    list->head = node;
    list->length++;
    printf("Node Appended!\n");
    return;
}

void appendAtEnd(struct LinkedList *list, struct Node *node) {
    // if the list is empty
    if (list->length == 0) {
        appendAtStart(list, node);
        return;
    }

    // if list is not empty
    // traverse the list
    struct Node *current = list->head;
    while (current->next != NULL) {
        current = current->next;
    }
    current->next = node;
    list->length++;
    printf("Node Appended!\n");
    return;
}

void insertAtPosition(struct LinkedList *list, struct Node *node, int pos) {
    // check out of bounds
    if (pos < 0 || pos > list->length) {
        printf("Position out of bounds\n");
        return;
    }

    // if pos 0 
    if (pos == 0) {
        appendAtStart(list, node);
        return;
    }

    // append at end if pos == len
    if (pos == list->length) {
        appendAtEnd(list, node);
        return;
    }

    // append at position if the list is not empty
    struct Node *current = list->head;
    int index = 0;
    while (index != pos - 1) {
        current = current->next;
        index++;
    }
    node->next = current->next;
    current->next = node;
    list->length++;
    printf("Node Appended\n");
    return;
}

void printlist(struct LinkedList *list) {
    // traverse and print the elements
    struct Node *current = list->head;
    while (current != NULL) {
        printf("%d -> ", current->data);
        current = current->next;
    }
    printf("NULL\n");
    return;
}

int main() {
    struct LinkedList *list = createList();
    struct Node *node = createNode(5);
    struct Node *node2 = createNode(10);
    struct Node *node3 = createNode(15);
    struct Node *node4 = createNode(20);
    struct Node *node5 = createNode(25);

    appendAtStart(list, node);
    appendAtStart(list, node2);
    appendAtEnd(list, node3);
    appendAtEnd(list, node4);
    insertAtPosition(list, node5, 20);

    printf("%d\n", list->length);
    printlist(list);

    return 0;

}